### PR TITLE
support json array in headertojson

### PR DESCRIPTION
**Pull request by suchen-sci**

### CODE REVIEW

**Review by zouyingjie — APPROVED** [review]

**Review by samutamm — APPROVED** [review]

**Comment by suchen-sci** [comment]
@localvar updated!

File: pkg/filter/headertojson/headertojson.go [hunk 1]
```diff
@@ -129,17 +193,26 @@ func (h *HeaderToJSON) handle(ctx context.HTTPContext) string {
...
-		return resultJSONEncodeDecodeErr
+		return resultBodyReadErr
 	}
-	for k, v := range headerMap {
-		bodyMap[k] = v
+
+	var body interface{}
+	if len(reqBody) == 0 {
+		m := make(map[string]interface{})
+		body = addHeaderToMap(m, headerMap)
```
Comment (localvar) [inline pkg/filter/headertojson/headertojson.go:204]:
```suggestion
		body = headerMap
```

File: pkg/filter/headertojson/headertojson.go [hunk 2]
```diff
@@ -97,24 +103,82 @@ func (h *HeaderToJSON) Status() interface{} {
...
+}
+
+func addHeaderToArray(bodyArray []map[string]interface{}, headerMap map[string]interface{}) interface{} {
+	for i := range bodyArray {
+		for k, v := range headerMap {
+			bodyArray[i][k] = v
+		}
+	}
+	return bodyArray
+}
```
Comment (localvar) [inline pkg/filter/headertojson/headertojson.go:144]:
```suggestion
func mergeMap(m1 map[string]interface{}, m2 map[string]interface{}) map[string]interface{} {
	for k, v := range m2{
		m1[k] = v
	}
	return m1
}
```

File: pkg/filter/headertojson/headertojson.go [hunk 3]
```diff
@@ -97,24 +103,82 @@ func (h *HeaderToJSON) Status() interface{} {
...
+		return addHeaderToMap(bodyMap, headerMap), nil
+
+	} else if char == '[' {
+		bodyArray, err := decodeArrayJSON(reqBody)
+		if err != nil {
+			return nil, errJSONEncodeDecode
+		}
+		return addHeaderToArray(bodyArray, headerMap), nil
+	}
+	return nil, errJSONEncodeDecode
```
Comment (localvar) [inline pkg/filter/headertojson/headertojson.go:181]:
```suggestion
func firstNonBlankByte(bytes []byte) byte {
	for _, b := range bytes {
		switch b {
		case '\n', '\t', '\r', ' ':
			continue
		}
		return b
	}
	return 0
}

func getNewBody(reqBody []byte, headerMap map[string]interface{}) (interface{}, error) {
	char := firstNonBlankByte(reqBody)

	if char == '{' {
		bodyMap, err := decodeMapJSON(reqBody)
		if err != nil {
			return nil, errJSONEncodeDecode
		}
		return mergeMap(bodyMap, headerMap), nil
	}
	
	if char == '[' {
		bodyArray, err := decodeArrayJSON(reqBody)
		if err != nil {
			return nil, errJSONEncodeDecode
		}
		for i := range bodyArray {
		    bodyArray[i] = mergeMap(bodyArray[i], headerMap)
		}
		return bodyArray, nil
	}
	
	return nil, errJSONEncodeDecode
```

Comment (suchen-sci) [reply pkg/filter/headertojson/headertojson.go]: updated! thanks a lot for your review! learn a lot from it!

**Review by localvar — APPROVED** [review]

**Pull request merged**